### ISSUE TITLE
bug: SecretsManager ClientRequestToken Ignored on PutSecretValue

### ISSUE THREAD

**Issue by c1moore** [localstack/localstack#5299]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When creating a new version of a secret, the `ClientRequestToken` is not used as the `VersionId`.  It appears localstack generates a new `VersionId` every time the API is hit.

### Expected Behavior

> A unique identifier for the new version of the secret. If you use the Amazon Web Services CLI or one of the Amazon Web Services SDKs to call this operation, then you can leave this parameter empty because they generate a random UUID for you. If you don't use the SDK and instead generate a raw HTTP request to the Secrets Manager service endpoint, then you must generate a ClientRequestToken yourself for new versions and include that value in the request. This value helps ensure idempotency. Secrets Manager uses this value to prevent the accidental creation of duplicate versions if there are failures and retries during the Lambda rotation function processing. We recommend that you generate a UUID-type value to ensure uniqueness within the specified secret. If the ClientRequestToken value isn't already associated with a version of the secret then a new version of the secret is created. If a version with this value already exists and that version's SecretString or SecretBinary values are the same as those in the request then the request is ignored. The operation is idempotent. If a version with this value already exists and the version of the SecretString and SecretBinary values are different from those in the request, then the request fails because you can't modify a secret version. You can only create new versions to store new secret values.  _**This value becomes the VersionId of the new version.**_

Bold and italics added for emphasis.  I left the long description because I think the behavior when a duplicate `VersionId` is seen is also important.

### How are you starting LocalStack?

With a docker-compose file

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

```yml
version: "3.1"

services:
  localstack:
    image: localstack/localstack:latest
    ports:
      - 4566:4566
```

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

```typescript
await this.secretsManager.putSecretValue({
  SecretId: event.SecretId,
  ClientRequestToken: event.ClientRequestToken,
  SecretString: mySecret,
  VersionStages: [SecretStageLabel.Pending],
}).promise();
```

### Environment

```markdown
- OS: Ubuntu 20.04
- LocalStack: latest
```

**Comment by c1moore** [comment]
There may be a similarly related bug where when creating a new secret value with the `AWSPENDING` stage label, it removes the `AWSCURRENT` stage label from the current version, which isn't supposed to happen until you issue the `UpdateSecretVersionStage` command.

**Comment by c1moore** [comment]
What's the status of this, it seems like it got closed and reopened?

To add to this a little more,  I noticed another issue.  The `ClientRquestToken` appears to always be the same, whether I invoke the rotation in code via

```ts
await this.secretsManager.rotateSecret({
  SecretId: secretId,
  RotationLambdaARN: this.lambdaArn,
  RotationRules: {
    AutomaticallyAfterDays: rotationRate,
  },
}).promise();
```

or in the console with

```bash
awslocal secretsmanager rotate-secret --secret-id /secret/id
```

For this secret, I save the `ClientRequestToken` in a JSON blob.  When I run the above command, I see the CLI output a different VersionId each time, but when I look at the current secret value with 

```bash
awslocal secretsmanager get-secret-value --secret-id /secret/id --version-stage AWSCURRENT
```

The `ClientRequestToken` field I save in the JSON blob never changes.

I'm wondering if this is the root of the issue.  I've verified this functionality works as expected in AWS, so it shouldn't be something odd I'm doing.

**Comment by MEPalma** [comment]
Omitting a `ClientRequestToken` value in `PutSecretValue` request is not the direct source of the issues you've been experiencing. Some test replicating your description have been investigated as you can see here: https://github.com/localstack/localstack/pull/5545.

However, the second observation, about using `PutSecretValue` to update the `stage labels`, did in fact have some issues. These have been fixed, and you can find some test scenarios in the same PR https://github.com/localstack/localstack/pull/5545.

Could you please confirm this new issue now concerning rotateSecret, you are able to reproduce on the latest version of LocalStack?

**Comment by c1moore** [comment]
I repulled the latest image yesterday right before I added the above comment.  It looks like now the `RemoveFromVersionId` field is being ignored.  Here's the result of DescribeSecret after UpdateSecretVersionStage:

```json
"VersionIdsToStages": {
   "a4651c21-055e-49b9-8eb1-f7a4fba5905c": [
       "AWSPREVIOUS"
   ],
   "4e34025e-8789-431a-8d92-93cf76d31401": [
       "AWSPENDING",
       "AWSCURRENT"
   ],
   "6bb0c339-4239-4728-ba36-027e229ab708": [
       "AWSPENDING",
       "AWSPREVIOUS"
   ]
 }
```

This was the command I ran

```bash
awslocal secretsmanager update-secret-version-stage --secret-id /cas/iam/jwt/c56fa25d-bfee-4a67-9a98-5900bf46a71a/key-pair --version-stage AWSCURRENT --move-to-version-id 4e34025e-8789-431a-8d92-93cf76d31401 --remove-from-version-id 6bb0c339-4239-4728-ba36-027e229ab708
```

> Omitting a ClientRequestToken value in PutSecretValue request is not the direct source of the issues you've been experiencing.

The issue as I understood it wasn't that I was omitting the `ClientRequestToken`.  According to the docs, if it's omitted, a unique one will be generated.  However, what I'm observing is that the same `ClientRequestToken` is being used every time the secret is generated.  For example, yesterday I rotated a secret 4 times (not including creating it).  Every time I rotated it, the field in the JSON blob I set based on the `ClientRequestToken` was always `c0c62618-b850-4a33-a553-25a60933842e`.  In production, when I rotate the secret, I see this value change.

I can open a separate ticket for that, though.

**Comment by c1moore** [comment]
The API endpoint for updating the version stage also appears to return the wrong response now.  I'm now getting this error message

```
SyntaxError: Unexpected token / in JSON at position 0
    at JSON.parse (<anonymous>)
    at JsonParser.parse (./node_modules/aws-sdk/lib/json/parser.js:6:25)
    at Request.extractData (./node_modules/aws-sdk/lib/protocol/json.js:63:24)
    at Request.callListeners (./node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (./node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (./node_modules/aws-sdk/lib/request.js:686:14)
    at Request.transition (./node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (./node_modules/aws-sdk/lib/state_machine.js:14:12)
    at ./node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (./node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (./node_modules/aws-sdk/lib/request.js:688:12)
    at Request.callListeners (./node_modules/aws-sdk/lib/sequential_executor.js:116:18)
    at Request.emit (./node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (./node_modules/aws-sdk/lib/request.js:686:14)
    at Request.transition (./node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (./node_modules/aws-sdk/lib/state_machine.js:14:12)
    at ./node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (./node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (./node_modules/aws-sdk/lib/request.js:688:12)
    at Request.callListeners (./node_modules/aws-sdk/lib/sequential_executor.js:116:18)
    at callNextListener (./node_modules/aws-sdk/lib/sequential_executor.js:96:12)
    at IncomingMessage.onEnd (./node_modules/aws-sdk/lib/event_listeners.js:335:13)
    at IncomingMessage.emit (node:events:402:35)
    at IncomingMessage.emit (node:domain:475:12)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
```

I hit the API directly from Postman just to confirm it's returning the wrong type and can confirm.  It should be returning a JSON object, but it's returning a string (the secret ID) instead.

```
POST / HTTP/1.1
Host: localhost:4566
X-Amz-Target: secretsmanager.UpdateSecretVersionStage
Content-Type: application/x-amz-json-1.1
Accept-Encoding: identity
x-amz-date: 20220305T205748Z
X-Amz-Content-Sha256: beaead3198f7da1e70d03ab969765e0821b24fc913697e929e726aeaebf0eba3
Authorization: AWS4-HMAC-SHA256 Credential=foo/20220305/us-east-1/secretsmanager/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=2054116ac1ad937eb71599efb1cd65f31a37ac8fe8768b5f92a2f84010b56fd4
Content-Length: 244

{
    "SecretId": "/my/secret",
    "VersionStage": "AWSCURRENT",
    "RemoveFromVersionId": "ebd0e679-3ad9-445f-bcc2-27868aef4f2f",
    "MoveToVersionId": "e130a93f-0816-41c1-b330-9ab3cc988a2e"
}
```

The response:

```
/my/secret
```

Expected response (from the docs):

```
{
  "ARN":"arn:aws:secretsmanager:us-west-2:123456789012:secret:MyTestDatabaseSecret-a1b2c3",
  "Name":"MyTestDatabaseSecret"
}
```

**Comment by c1moore** [comment]
Has there been any progress on this?  To summarize the errors as I see them right now

1. `UpdateSecretVersionStage` is returning an invalid response (returning a string instead of the expected JSON)
2. `ClientRequestToken` does not appear to be unique when generated by localstack.  This was tested using the Node AWS SDK and the AWS CLI.
3. `AWSPENDING` is not removed when the secret version stage is updated to `AWSCURRENT`.  This is something AWS does automatically and does not require a separate request.
4. localstack does not remove the `AWSPREVIOUS` version stage from the version that was previously marked as `AWSPREVIOUS` when changing the `AWSCURRENT` tag.

This is completely blocking me at this point because SecretsManager is completely unusable in localstack.

**Comment by MEPalma** [comment]
Thank you for the patience and providing the summary above. These new points will soon get investigated.

**Comment by c1moore** [comment]
I pulled the latest changes and at least point 1 above appears to be addressed, but point 2 is still an issue.  The issues with the version stages may also be addressed, but I haven't tested extensively.

**Comment by c1moore** [comment]
@MEPalma It's been almost 2 months since this problem was reported and the original issue is still the issue I'm running into.  I tried to get around the issue of localstack reusing the same `ClientRequestToken` by specifying the `ClientRequestToken` when making the request to rotate the secret, but localstack seems to ignore the one provided (probably harkens back to the original report that " the `ClientRequestToken` is not used as the `VersionId`").

I've tried scanning through the code, but I can't find the actual implementation.  I'm not sure if it's open source, but if it is, could you point me to the right place?  I wouldn't mind contributing to the project to get unblocked.

**Comment by MEPalma** [comment]
With regards to the initial statement concerning the ClientRequestToken in PutSecretValue requests, this was fixed, and http tests following your description were added. Could you give me some feedback in those? They were crosschecked against AWS.

Concerning the latest descriptions, now pointing to version stages being corrupted, this is true, and is in the process of being fixed. It is unrelated to the ClientRequestToken missing, hence why you're probably still experiencing some issues. 
I would suggest waiting for the version stages update logic to be fixed, or attempt to better describe the workflow in which you do see the clientRequestToken being as issue.

**Comment by MEPalma** [comment]
Here is the fixed I have made about the ClientRequestToken following your initial description. Please do let me know if you disagree with something.

I am sorry this issue is still not resolved for you.

Fix: https://github.com/spulec/moto/pull/4851

**Comment by c1moore** [comment]
@MEPalma It looks like you fixed the issue for `PutSecret`, but I'm still having an issue with `RotateSecret`.

I think the issue I'm having is due to [these lines](https://github.com/MEPalma/moto/blob/cd6cff5c966a93cffb71dd90b49ebedb8785a93d/moto/secretsmanager/models.py#L504-L514).  That's not part of the AWS standards.  The first step of `RotateSecret` (`createSecret`) is supposed to `PutSecretValue` and attach the `AWSPENDING` stage to it (basically what these lines are doing).  AWS also suggests checking if the secret version already exists before creating the secret to avoid scenarios where AWS could call `createSecret` multiple times even though the new secret version was created.  An example algorithm of how to properly write a lambda for secret rotation is provided by AWS [here](https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerRotationTemplate/lambda_function.py#L72-L104).

My code pretty much follows that same algorithm.  So effectively I'm getting to [this line](https://github.com/aws-samples/aws-secrets-manager-rotation-lambdas/blob/252449551d58075f353444743e81a8c56d9f96db/SecretsManagerRotationTemplate/lambda_function.py#L94), seeing that the secret already exists, and not generating the new secret value, which is an issue because of [this line](https://github.com/MEPalma/moto/blob/cd6cff5c966a93cffb71dd90b49ebedb8785a93d/moto/secretsmanager/models.py#L509).

It's not that localstack is reusing a UUID or ignoring the `ClientRequestToken` I'm providing, it's that localstack is preemptively generating the secret (performing the first step of secret rotation), by duplicating the original secret, which is a huge issue for my use case since these secrets must be unique.

**Comment by c1moore** [comment]
I'd recommend removing those lines completely.

**Comment by c1moore** [comment]
It took me a little while to find that and I thought I was loosing my mind, lol.  I'm glad that was helpful.

Any idea on what turnaround could look like?  I know nothing is guaranteed and it can be hard to estimate, but is it a couple of days/weeks/months?

**Comment by MEPalma** [comment]
I'd be surprised if it took more than a week; that being said it depends on how it'll sync with release dates. I'll keep you updated.
